Keras inference time optimizer

ZFTurbo opened the pull request:
Reduce neural net structure (Conv + BN -> Conv)
Also works:
DepthwiseConv2D + BN -> DepthwiseConv2D
SeparableConv2D + BN -> SeparableConv2D

This code takes on input trained Keras model and optimize layer structure and weights in such a way that model became much faster (~30%), but works identically to initial model. It can be extremely useful in case you need to process large amount of images with trained model. Reduce operation was tested on all Keras models zoo. See comparison table and full description by link:
https://github.com/ZFTurbo/Keras-inference-time-optimizer

ahundt commented:
The style here doesn't seem to match that of other keras code. Could you fix that?

titu1994 commented:
Are there no tests other than the benchmark test from your repo that can be performed ? A benchmark test isn't what would be feasible inside Travis, especially with such large models.

Once the style fixes are finished, this can be merged if it includes at least some form of test and all tests pass.

ZFTurbo commented:
@titu1994 I added "mobilenet_small" benchmark. It run several seconds on my machine. I think it can be used as test for kito.py code. 

Should I include tests directly in kito.py file?

@ahundt: Regarding style. I saw other files in "utils" folder and didn't see any specific style pattern. Could you please clarify what exactly should I change? Do you mean comments in functions or something else?

ahundt commented:
tests should go in the tests folder, take a look how those are written.

titu1994 commented:
@ZFTurbo The test is short on Tensorflow, but I doubt its fast on Theano. In any case, it's better than no test at all.

As @ahundt mentioned, tests should go into the `tests` directory, specifically `tests/keras_contrib/utils/*_test.py` where you can replace * with some name. For the format, please see other tests or tests from core Keras.

ZFTurbo commented:
I added test.

ahundt commented:
@titu1994 @ZFTurbo this looks pretty good to me now, think it should be merged?

ZFTurbo commented:
@ahundt I'm totally fine with merge. )

ahundt commented on keras_contrib/utils/kito.py:
```diff
@@ -0,0 +1,309 @@
+
+    # DeepLabV3+ non-standard layer
+    if layer.__class__.__name__ == 'BilinearUpsampling':
+        from neural_nets.deeplab_v3_plus_model import BilinearUpsampling
```
Just noticed this, there is now a bilinear upsampling layer in keras itself.
https://github.com/keras-team/keras/pull/10994

ZFTurbo replied:
It has other name in Keras. This code made specially for DeepLab V3+. It still uses it's own implementation.

bhack replied:
/cc @bonlime

ahundt replied:
We can't have specialized code for things not included in keras-contrib. You'll need to come up with a way to parameterize this, such as by passing functions, or a dictionary.

ahundt commented on tests/keras_contrib/utils/kito_test.py:
```diff
@@ -0,0 +1,121 @@
+        print('Reduced model number layers: {}'.format(len(model_reduced.layers)))
+        print('Compare models...')
+        if model_name in ['nasnetlarge', 'deeplab_v3plus_mobile', 'deeplab_v3plus_xception']:
+            max_error = compare_two_models_results(model, model_reduced, test_number=10000, max_batch=128)
```
how long does this test take to run?

ZFTurbo replied:
Several seconds

ahundt commented on keras_contrib/utils/kito.py:
```diff
@@ -0,0 +1,309 @@
+    return outbound_layers
+
+
+def get_copy_of_layer(layer, verbose=False):
```
how do users do this for their own non-standard layers?

ZFTurbo replied:
They need to change the code ( Like here:

    # DeepLabV3+ non-standard layer
    if layer.__class__.__name__ == 'BilinearUpsampling':
        from neural_nets.deeplab_v3_plus_model import BilinearUpsampling

I'm not sure how to make it universal, since parameters are different.

ahundt replied:
You can also have a look at keras itself registers functions. If you look at the python file for loss, at the very bottom there is some registration code.

ZFTurbo commented:
1) Added more tests
2) Removed non-standard layers. Probably need to add custom_objects parameter. Will do it later.
3) Added support for layers of type "Model".

gabrieldemarmiesse commented:
Thank you @ZFTurbo for your work.

I've looked at the result presented in your page and the number are impressive.
Since you're using tensorflow, I might add that there is XLA (maybe you already know about it) which does a similar job.

If it's not too much trouble, can I ask you to run your benchmarks with XLA enabled?

Also as a side note. I started using keras for the first time with your scripts on kaggle a few years ago. Thanks for sharing! (remember the fish competition?) :)

ZFTurbo commented:
> 
> 
> Thank you @ZFTurbo for your work.
> 
> I've looked at the result presented in your page and the number are impressive.
> Since you're using tensorflow, I might add that there is XLA (maybe you already know about it) which does a similar job.
> 
> If it's not too much trouble, can I ask you to run your benchmarks with XLA enabled?

I heard about XLA but not tried it yet. Probably XLA will make KITO useless. I'll try it and post result here.

> Also as a side note. I started using keras for the first time with your scripts on kaggle a few years ago. Thanks for sharing! (remember the fish competition?) :)

Good to know it )) That's bad, but looks like PyTorch becoming mainstream now, instead of Keras.
